src/ck-ldb/TreeLevel.h

```diff
@@ -0,0 +1,1236 @@
+          tokens.emplace_back(local_id, oldPe, oload);
+#if DEBUG__TREE_LB_L2
+          CkPrintf("[%d] Sending obj with local_obj_id=%d oldPe=%d load=%f TO %d\n",
+                   CkMyPe(), local_id, oldPe, oload, dest);
+#endif
+        }
+        int ntokens = tokens.size();
+        TokenListMsg* token_set_msg = new (ntokens, ntokens, ntokens, 0) TokenListMsg;
+        token_set_msg->load = nominal_load;
+        token_set_msg->num_tokens = ntokens;
+        for (int j = 0; j < ntokens; j++)
+        {
+          auto& token = tokens[j];
+          token_set_msg->local_ids[j] = token.obj_local_id;
+          token_set_msg->oldPes[j] = token.oldPe;
+          token_set_msg->loads[j] = token.load;
+        }
+        token_sets.push_back(token_set_msg);
+      }
+    }
+    return outgoing_nominal_load;
+  }
+
+  virtual int tokensReceived(TreeLBMessage* msg)
+  {
+    IStrategyWrapper* wrapper = wrappers[current_strategy];
+    TokenListMsg* token_set = (TokenListMsg*)msg;
+    for (int i = 0; i < token_set->num_tokens; i++)
+    {
+#if DEBUG__TREE_LB_L2
+      CkPrintf("[%d] Adding object with local_id=%d from oldPe=%d with load %f\n",
+               CkMyPe(), token_set->local_ids[i], token_set->oldPes[i],
+               token_set->loads[i]);
+#endif
+      wrapper->addForeignObject(token_set->local_ids[i], token_set->oldPes[i],
+                                token_set->loads[i]);
+    }
+    int load = token_set->load;
+#if DEBUG__TREE_LB_L2
+    CkPrintf("[%d] Total nominal load in token set is %d\n", CkMyPe(), load);
+#endif
+    delete token_set;
+    return load;
+  }
+
+  virtual void loadBalance(std::vector<TreeLBMessage*>& decisions, IDM& idm)
+  {
+    CkAssert(wrappers.size() > current_strategy);
+    IStrategyWrapper* wrapper = wrappers[current_strategy];
+    CkAssert(wrapper != nullptr);
+
+    if (cutoff())
+    {
+      num_children = stats_msgs.size();
+      CkAssert(num_children > 0);
+#if DEBUG__TREE_LB_L2
+      CkPrintf("[%d] NodeSetLevel::loadBalance (w cutoff), num_children=%d m=%d n=%d\n",
+               CkMyPe(), num_children, m, n);
+#endif
+
+      CkAssert(migMsg == nullptr);
+      int npes = CkNumPes();
+      migMsg = new (npes, npes, n, 0) LLBMigrateMsg;
+      migMsg->n_moves = 0;
+      std::fill(migMsg->num_incoming, migMsg->num_incoming + npes, 0);
+
+      wrapper->prepStrategy(n, m, stats_msgs, migMsg);
+      // need to cast pointer to ensure delete of CMessage_LBStatsMsg_1 is called
+      for (auto msg : stats_msgs) delete (LBStatsMsg_1*)msg;
+      stats_msgs.clear();
+      m = n = 0;
+    }
+    wrapper->runStrategy(migMsg, &idm);
+    if (current_strategy == wrappers.size() - 1)
+    {
+      if (repeat_strategies) current_strategy = 0;
+    }
+    else
+    {
+      current_strategy++;
+    }
+    decisions.resize(num_children);
+    decisions[0] = migMsg;
+    for (int i = 1; i < num_children; i++)
+      decisions[i] = (TreeLBMessage*)CkCopyMsg((void**)&migMsg);
```

Comment:
Could this be CmiReference instead of CkCopyMsg?

Reply:
I don't think there's a good way to do this with Charm messages. Things get wonky if you try to send them more than once.

Do you know if there's a way to get that to work with `CmiReference`? Also, note that these are variable sized messages, so they need to be allocated in a specific way.

Reply:
I think you can do something like 'CmiReference(UsrToEnv(mig));' but it might not work in this case.

Reply:
I'll leave this as future work, I've been trying to implement it over the past week and it's been full of thorns, it's a bit complicated because when the messages are sent, we can't assume that they're all the same message when they're sent because (while it's not implemented yet), some parts of the infrastructure may create different messages for each of the children, so using section multicast or multicast converse sends won't work. That coupled with actually making sure the messages get freed has complicated things, causing ref counting errors in CmiFree. I think some scheme with `CmiReference` can probably work, but I don't think it should stop get in the way of merging this. I've made #2857 for this.